### ISSUE TITLE
applying toll to public transport

### ISSUE THREAD

**Issue by SarahSie** [matsim-org/matsim-code-examples#613]
Hello,

I would like to know how I can apply toll to the public transport in MATSim. I could not find relevant information in the MATSim book.
It would be great if you can help me,

Kind regards,
S

**Comment by JWJoubert** [comment]
Hi @CharleneSie, I would suggest that you use the `roadpricing` contrib, and then specifically implement a `TollFactor`. That class has a single required method to be implemented, namely
```
double getTollFactor(Id<Person> personId, Id<Vehicle> vehicleId, Id<Link> linkId, double time);
``` 
And since public transport vehicles have a unique ` Id<Vehicle>`, you can apply the toll to them (only), and disregard the rest (or simply return `0.0` for the rest).

**Comment by SarahSie** [comment]
Many thanks, @JWJoubert, for your reply and precise explanation. that's a smart way, however, I aim to let the controller compute the toll and evaluate scores including tolls. How can I do this in the scoring?

**Comment by JWJoubert** [comment]
> I aim to let the controller compute the toll and evaluate scores including tolls.

Exactly, that is why you need the `roadpricing` contrib. It takes care of that. Have a look at the example in `org.matsim.contrib.roadpricing.run.RunRoadPricingUsingTollFactorExample`.

Instead of lines 61-69, I suggest you create your own class implementation of `TollFactor`. It is likely going to be a bit more readable than the example class above. But then again, the example class is just there as a bare-bones minimalist example.

**Comment by vsp-gleich** [comment]
Hello,
I fear that there is some confusion on tolls and scores in this issue. The above should lead to applying a toll to the driver of the pt vehicle. However, that driver has no score and no replanning logic, he simply follows the transit schedule. So there will be no behaviour reaction to the toll, only a few more PersonMoneyEvents in the output events file that could be used for analysis after the simulation has finished.
This issue reminds me of the recent discussion on minibus and roadpricing: [https://github.com/matsim-org/matsim-libs/issues/1581](https://github.com/matsim-org/matsim-libs/issues/1581)

**Comment by JWJoubert** [comment]
Oh, golly, you're 100% correct, @vsp-gleich.

 @CharleneSie, sorry, but I forgot that we actually implemented quite a specific variant on our end to get that done (in the `minibus` contribution) and get behavioural sensitivity. And, if I'm not mistaken, @vsp-gleich, that pull request did go through... but that is then only applicable to the minibus/paratransit, and _not formal transit_.
